src/main/scala/viper/server/frontends/lsp/FileManager.scala

```diff
@@ -0,0 +1,382 @@
+import viper.server.frontends.lsp
+import viper.server.frontends.lsp.VerificationState._
+import viper.server.frontends.lsp.VerificationSuccess._
+import viper.server.vsi.JobID
+import viper.silver.ast.{Domain, Field, Function, Method, Predicate, SourcePosition}
+import viper.silver.reporter._
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable.ArrayBuffer
+
+class FileManager(file_uri: String) {
+  // file under verification
+  var uri: URI = URI.create(file_uri)
+  var path: Path = Paths.get(uri)
+  var filename: String = path.getFileName.toString
+  var lastSuccess: VerificationSuccess = NA
+  var internalErrorMessage: String = ""
+
+  //state specific to one verification
+  var is_aborting: Boolean = false
+  var is_verifying: Boolean = false
+  var global_failure: Boolean = false
+  var state: VerificationState = Stopped
+  var manuallyTriggered: Boolean = _
+
+  //verification results
+  var jid: Int = -1
+  var time: Long = 0
+  var diagnostics: ArrayBuffer[Diagnostic] = _
+//  var verifiables: Array[Verifiable] = _
+  var parsingCompleted: Boolean = false
+  var typeCheckingCompleted: Boolean = false
+  var backendType: String = _
+  var progress: Progress = _
+//  var shownExecutionTrace: Array[ExecutionTrace] = _
+  var symbolInformation: ArrayBuffer[SymbolInformation] = ArrayBuffer()
+  var definitions: ArrayBuffer[lsp.Definition] = ArrayBuffer()
+
+  //working variables
+  private var lines: Array[String] = Array()
+  private var wrongFormat: Boolean = false
+  private var partialData: String = ""
+
+  def resetDiagnostics(): Unit = {
+    diagnostics = ArrayBuffer()
+    val diagnosticParams = new PublishDiagnosticsParams()
+    diagnosticParams.setUri(file_uri)
+    diagnosticParams.setDiagnostics(diagnostics.asJava)
+    Coordinator.client.publishDiagnostics(diagnosticParams)
+  }
+
+  def resetLastSuccess(): Unit = {
+    lastSuccess = NA
+  }
+
+  def prepareVerification(): Unit = {
+    is_verifying = true
+    is_aborting = false
+    state = Stopped
+    lines = Array()
+    wrongFormat = false
+    if (partialData.length > 0) {
+      Log.debug("Some unparsed output was detected:\n" + partialData)
+      partialData = ""
+    }
+    time = 0
+    resetDiagnostics()
+//    verifiables = Array()
+    parsingCompleted = true
+    typeCheckingCompleted = true
+    internalErrorMessage = ""
+  }
+
+  def abortVerification(): CFuture[Void] = {
+    if (!is_verifying) {
+      return CFuture.completedFuture(null)
+    }
+    Log.info("Aborting running verification.")
+    is_aborting = true
+    Coordinator.verifier.stopVerification(JobID(jid)).thenAccept(_ => {
+      is_verifying = false
+      lastSuccess = Aborted
+    }).exceptionally(e => {
+      Log.debug("Error aborting verification of " + filename + ": " + e)
+      null
+    })
+  }
+
+  def startStageProcess(fileToVerify: String): Option[String] = {
```

Comment:
Ideally, we do not need the concept of stages anymore as it is obsolete. Can we easily carve out all the code that is related to this feature?

Reply:
Yes. this is something that should reworked in conjunction with the settings. I have now removed all the code that relates to this concept.